### PR TITLE
[20335] DynamicTypeBuilderFactoryImpl::create_type_w_type_object implementation

### PR DESCRIPTION
**Pull request by JLBuenoLopez-eProsima**
## Description

This PR implements `DynamicTypeBuilderFactoryImpl::create_type_w_type_object` method. It also uses this feature to fix the DDS SQL Filter and its tests.

## Contributor Checklist
- [ ] Commit messages follow the project guidelines. 
- [ ] The code follows the style guidelines of this project. 
- [ ] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [ ] Any new/modified methods have been properly documented using Doxygen. 
- [ ] Changes are ABI compatible. 
- [ ] Changes are API compatible. 
- [ ] New feature has been added to the `versions.md` file (if applicable).
- [ ] New feature has been documented/Current behavior is correctly described in the documentation. 
    
- [ ] Applicable backports have been included in the description.

## Reviewer Checklist
- [ ] The PR has a milestone assigned.
- [ ] The title and description correctly express the PR's purpose.
- [ ] Check contributor checklist is correct.
- [ ] Check CI results: changes do not issue any warning.
- [ ] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: src/cpp/fastdds/xtypes/dynamic_types/DynamicTypeBuilderFactoryImpl.cpp [hunk 1]
```diff
@@ -259,6 +371,1171 @@ ReturnCode_t DynamicTypeBuilderFactoryImpl::delete_type(
...
+            ret_val = get_primitive_type(TK_FLOAT128);
+            break;
+        case xtypes::TK_CHAR8:
+            ret_val = get_primitive_type(TK_CHAR8);
+            break;
+        case xtypes::TK_CHAR16:
+            ret_val = get_primitive_type(TK_CHAR16);
+            break;
+        case xtypes::TI_STRING8_SMALL:
+            ret_val = create_string_type(type_identifier.string_sdefn().bound())->build();
```
Comment (richiware) [inline src/cpp/fastdds/xtypes/dynamic_types/DynamicTypeBuilderFactoryImpl.cpp:1277]: Detect bound is zero and set LENGTH_UNLIMITED.

File: src/cpp/fastdds/xtypes/dynamic_types/DynamicTypeBuilderFactoryImpl.cpp [hunk 2]
```diff
@@ -259,6 +371,1171 @@ ReturnCode_t DynamicTypeBuilderFactoryImpl::delete_type(
...
+            break;
+        case xtypes::TI_STRING16_SMALL:
+            ret_val = create_wstring_type(type_identifier.string_sdefn().bound())->build();
+            break;
+        case xtypes::TI_STRING8_LARGE:
+            ret_val = create_string_type(type_identifier.string_ldefn().bound())->build();
```
Comment (richiware) [inline src/cpp/fastdds/xtypes/dynamic_types/DynamicTypeBuilderFactoryImpl.cpp:1283]: Detect bound is zero and set LENGTH_UNLIMITED.

File: src/cpp/fastdds/xtypes/dynamic_types/DynamicTypeBuilderFactoryImpl.cpp [hunk 3]
```diff
@@ -259,6 +371,1171 @@ ReturnCode_t DynamicTypeBuilderFactoryImpl::delete_type(
...
+        case xtypes::TI_PLAIN_SEQUENCE_LARGE:
+        {
+            traits<DynamicType>::ref_type element_type = base_type_from_type_identifier(
+                    *type_identifier.seq_ldefn().element_identifier());
+            ret_val = create_sequence_type(element_type, type_identifier.seq_ldefn().bound())->build();
+            break;
+        }
+        case xtypes::TI_PLAIN_ARRAY_SMALL:
+        {
+            TypeDescriptorImpl small_array_descriptor {TK_ARRAY, ""};
```
Comment (richiware) [inline src/cpp/fastdds/xtypes/dynamic_types/DynamicTypeBuilderFactoryImpl.cpp:1306]: Why don't you use `create_array_type()`.

File: src/cpp/fastdds/xtypes/dynamic_types/DynamicTypeBuilderFactoryImpl.cpp [hunk 4]
```diff
@@ -259,6 +371,1171 @@ ReturnCode_t DynamicTypeBuilderFactoryImpl::delete_type(
...
+        for (const xtypes::AppliedAnnotation& annotation : ann_custom.value())
+        {
+            // Find annotation type object
+            xtypes::TypeObject annotation_type_object;
+            if (RETCODE_OK !=
+                    fastrtps::rtps::RTPSDomainImpl::get_instance()->type_object_registry_observer().get_type_object(
+                    annotation.annotation_typeid(), annotation_type_object))
+            {
+                EPROSIMA_LOG_ERROR(DYN_TYPES, "Given annotation type identifier unknown to TypeObjectRegistry");
+                ret_val.reset();
```
Comment (richiware) [inline src/cpp/fastdds/xtypes/dynamic_types/DynamicTypeBuilderFactoryImpl.cpp:1387]: break?

File: src/cpp/fastdds/xtypes/dynamic_types/DynamicTypeBuilderFactoryImpl.cpp [hunk 5]
```diff
@@ -259,6 +371,1171 @@ ReturnCode_t DynamicTypeBuilderFactoryImpl::delete_type(
...
+        MemberDescriptor::_ref_type& ret_val,
+        const xtypes::MemberFlag& flags)
+{
+    if (flags & xtypes::TRY_CONSTRUCT2)
+    {
+        ret_val->try_construct_kind(TryConstructKind::USE_DEFAULT);
+    }
+    else if (flags & xtypes::TRY_CONSTRUCT1 && flags & xtypes::TRY_CONSTRUCT2)
+    {
+        ret_val->try_construct_kind(TryConstructKind::TRIM);
```
Comment (richiware) [inline src/cpp/fastdds/xtypes/dynamic_types/DynamicTypeBuilderFactoryImpl.cpp:1452]: This line will be never called due to the previous condition.

File: src/cpp/fastdds/xtypes/dynamic_types/DynamicTypeBuilderImpl.cpp [hunk 6]
```diff
@@ -192,6 +192,14 @@ bool DynamicTypeBuilderImpl::equals(
             }
         }
 
+        if (verbatim_.size() == impl->verbatim_.size())
+        {
+            for (size_t count {0}; ret_value && count << verbatim_.size(); ++count)
```
Comment (richiware) [inline src/cpp/fastdds/xtypes/dynamic_types/DynamicTypeBuilderImpl.cpp:197]:
```suggestion
            for (size_t count {0}; ret_value && count < verbatim_.size(); ++count)
```

**Review by richiware — CHANGES_REQUESTED** [review]

File: src/cpp/fastdds/topic/DDSSQLFilter/DDSFilterFactory.cpp [hunk 7]
```diff
@@ -53,13 +54,13 @@ namespace DDSSQLFilter {
 
 static ReturnCode_t transform_enum(
         std::shared_ptr<DDSFilterValue>& value,
-        const eprosima::fastdds::dds::xtypes::TypeIdentifier* type,
+        const std::shared_ptr<xtypes::TypeIdentifier> type,
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply src/cpp/fastdds/topic/DDSSQLFilter/DDSFilterFactory.cpp:57]: This parameter is only used as input and it is not being modified within the method.

**Comment by JLBuenoLopez-eProsima** [comment]
@richiprosima please test this

**Comment by JLBuenoLopez-eProsima** [comment]
@richiprosima please test this

**Comment by JLBuenoLopez-eProsima** [comment]
@richiprosima please test this

**Comment by JLBuenoLopez-eProsima** [comment]
@richiprosima please test this

**Comment by JLBuenoLopez-eProsima** [comment]
@richiprosima please test this

**Comment by JLBuenoLopez-eProsima** [comment]
@richiprosima please test this

**Review by richiware — CHANGES_REQUESTED** [review]

**Review by adriancampo — APPROVED** [review]
LGTM. Waiting for CI

File: src/cpp/fastdds/topic/DDSSQLFilter/DDSFilterFactory.cpp [hunk 8]
```diff
@@ -518,40 +522,49 @@ ReturnCode_t DDSFilterFactory::create_content_filter(
...
-                while (expr->parameters.size() < n_params)
-                {
-                    expr->parameters.emplace_back();
-                }
-                ExpressionParsingState state{ &type_object, filter_parameters, expr };
-                ret = convert_tree<DDSFilterCondition>(state, expr->root, *(node->children[0]));
-                if (RETCODE_OK == ret)
+                DynamicType::_ref_type dyn_type = DynamicTypeBuilderFactory::get_instance()->create_type_w_type_object(
+                    type_objects->complete_type_object)->build();
+                if (nullptr != dyn_type)
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply src/cpp/fastdds/topic/DDSSQLFilter/DDSFilterFactory.cpp]: @richiware I am going to apply this suggestion correctly. This suggestion just broke the CFT feature :sweat_smile: 

```suggestion
                if (dyn_type)
```

@adriancampo FYI

**Pull request merged**